src/main/java/org/snomed/snowstorm/core/data/services/ConceptService.java

```diff
@@ -805,4 +809,110 @@ public void addClauses(Set<String> conceptIds, Boolean active, BoolQueryBuilder
 			conceptQuery.must(termsQuery(SnomedComponent.Fields.ACTIVE, active));
 		}
 	}
+
+	private String getModuleId(String branchPath) {
+		Branch branch = branchService.findBranchOrThrow(branchPath, true);
+		final boolean isExtension = (branch.getMetadata() != null && !Strings.isEmpty(branch.getMetadata().getString(BranchMetadataKeys.DEPENDENCY_PACKAGE)));
+
+		return isExtension ? branch.getMetadata().getString(BranchMetadataKeys.DEFAULT_MODULE_ID) : Concepts.CORE_MODULE;
+	}
+
+	public ConceptMini donateConcept(ConceptMini concept, String sourceBranchPath, String destinationBranchPath, List<LanguageDialect> languageDialects, boolean includeDependencies) throws ServiceException {
+		String sourceModuleId = getModuleId(sourceBranchPath);
+		String destinationModuleId = getModuleId(destinationBranchPath);
+
+		List<ReferenceSetMember> referenceSetMembers = new ArrayList<>();
+
+		// Get axioms
+		List<ReferenceSetMember> axiomMembers = referenceSetMemberService.findMembers(
+				sourceBranchPath,
+				new MemberSearchRequest()
+						.active(true)
+						.isNullEffectiveTime(false)
+						.referenceSet(Concepts.OWL_AXIOM_REFERENCE_SET)
+						.referencedComponentId(concept.getConceptId()), LARGE_PAGE).getContent();
+
+		// Check that the target concept exists on the destination branch
+		for (ReferenceSetMember axiomMember : axiomMembers) {
+			try {
+				SAxiomRepresentation axiomRepresentation = axiomConversionService.convertAxiomMemberToAxiomRepresentation(axiomMember);
+				if (axiomRepresentation != null) {
+					Set<Relationship> relationships = axiomRepresentation.getLeftHandSideNamedConcept() != null
+							? axiomRepresentation.getRightHandSideRelationships()
+							: axiomRepresentation.getLeftHandSideRelationships();
+
+					for (Relationship relationship : relationships) {
+						String targetConceptId = relationship.getDestinationId();
+						Concept targetConcept = find(targetConceptId, sourceBranchPath);
+						if (targetConcept == null) {
+							throw new ServiceException("Axiom " + axiomMember.getId() + ": target concept " + targetConceptId + " does not exist on path " + sourceBranchPath);
+						}
+						if (includeDependencies && targetConcept.getModuleId().equals(sourceModuleId)) {
+							donateConcept(new ConceptMini(targetConcept, languageDialects), sourceBranchPath, destinationBranchPath, languageDialects, true);
```

Comment:
@astro-snail This works for simple dependencies but might fail when there are circular dependencies. Let me check whether these complex cases do exist or not.

Reply:
@astro-snail  I have checked and it is not possible to have circular dependencies. Please ignore this. Thanks.